File: 290.word-pattern.java
/*
 * @lc app=leetcode id=290 lang=java
 *
 * [290] Word Pattern
 */
class Solution {
    public boolean wordPattern(String pattern, String str) {
        if (pattern == null || str == null) return false;
        String[] sl = str.split(" ");
        if (sl.length != pattern.length()) return false;
        Map<Character, String> dict = new HashMap<>();
        Set<String> seem = new HashSet<>();
        for (int i = 0; i < pattern.length(); ++i) {
            String word = dict.get(pattern.charAt(i));
            if (word == null) {
                dict.put(pattern.charAt(i), sl[i]);
                if (seem.contains(sl[i])) return false;
                seem.add(sl[i]);
            }
            else if (!word.equals(sl[i])) return false;
        }
        return true;
    }
}

File: go/290.word-pattern.go
package main

import (
	"strings"
)

/*
 * @lc app=leetcode id=290 lang=golang
 *
 * [290] Word Pattern
 */

// @lc code=start

type void struct{}

var member void

func wordPattern(pattern string, str string) bool {
	var lookup [26]string
	strs := strings.Split(str, " ")
	if len(pattern) != len(strs) {
		return false
	}
	for i, s := range strs {
		li := int(pattern[i]) - int('a')
		if p := lookup[li]; p != "" {
			if p != s {
				return false
			}
		} else {
			lookup[li] = s
		}
	}
	set := make(map[string]void)
	for _, s := range lookup {
		if s != "" {
			if _, b := set[s]; b {
				return false
			} else {
				set[s] = member
			}
		}
	}
	return true
}

// @lc code=end
